### DISCUSSION TITLE
can I set a device vector in a loop

### DISCUSSION THREAD

**Discussion by ztdepztdep** [NVIDIA/cccl#1047] [Thrust]
I want to set the device vector like the c++ vector 
for(int i=0;i<100;i++) device_vector[i]=3.14.

Does it works in thrust?

**Comment by miscco** [comment]
I would advice against that and instead either use the suitable constructor or `thrust::fill` in case of a constant input

**Comment by alliepiper** [comment]
@miscco is correct, it's better to do a bulk write into a `device_vector` instead of assigning each element.

Assigning in a loop will work, but it will be very slow and inefficient, as each individual element will be copied from host -> device memory in separate transactions. These copies are very expensive, and minimizing them using bulk update methods will improve performance significantly.

If each element has the same value (as in the example) the `thrust::device_vector<float> data(100, 3.14f)` constructor or `thrust::fill(data.begin(), data.end(), 3.14f)` will work, as mentioned above.

If the data is more complex, it's more efficient to create a `thrust::host_vector` with the data and then copy that into the `thrust::device_vector`:

```
thrust::host_vector<float> h_data;
// fill h_data
thrust::device_vector<float> d_data = h_data; // copies all data at once
```
